### PR TITLE
Export a Markdown note as PDF

### LINKED ISSUE #2013: PDF under 'Export As..'
Currently, we can get PDF by abusing the print-to-pdf feature. 
I would love to have a PDF option under the 'Export as' menu option. 

Please tag this as a feature request.

### PR DESCRIPTION
**Pull request by AgentEpsilon**
## Description

Adds an "Export to PDF" option for Markdown notes. This feature behaves consistently with the other export options - however, due to its use of a hidden Electron window, it runs asynchronously. I've added support for output formatters to return Promises in `exportData.js` to support this feature.

Additionally, I have fixed `exportFolder.js` to actually wait for all notes to export. This wasn't a problem when note exporting was synchronous, but now it seems the extra wait time was causing tests to fail.

screen shot 2018-12-03 at 17 09 12

## Issue fixed

Fixes #2013.

## Type of changes

- :white_circle: Bug fix (Change that fixed an issue)
- :white_circle: Breaking change (Change that can cause existing functionality to change)
- :radio_button: Improvement (Change that improves the code. Maybe performance or development improvement)
- :radio_button: Feature (Change that adds new functionality)
- :white_circle: Documentation change (Change that modifies documentation. Maybe typo fixes)

## Checklist:

- :radio_button: My code follows [the project code style](docs/code_style.md)
- :white_circle: I have written test for my code and it has been tested
- :radio_button: All existing tests have been passed
- :radio_button: I have attached a screenshot/video to visualize my change if possible

### CODE REVIEW

**Comment by ZeroX-DG** [comment]
I can't export notes with images. Can you check?

**Comment by ZeroX-DG** [comment]
@AgentEpsilon ping!

File: browser/components/MarkdownPreview.js [hunk 1]
```diff
@@ -294,58 +295,75 @@ export default class MarkdownPreview extends React.Component {
...
+
+  handleSaveAsPdf () {
+    this.exportAsDocument('pdf', (noteContent, exportTasks) => {
+      const printout = new remote.BrowserWindow({show: false})
+      printout.loadURL('data:text/html;charset=UTF-8,' + this.htmlContentFormatter(noteContent, exportTasks))
+      return new Promise((resolve, reject) => {
+        printout.webContents.on('did-finish-load', () => {
+          printout.webContents.printToPDF({}, (err, data) => {
+            if (err) reject(err)
+            else resolve(data)
```
[the remark this answers is not in the document]
Comment (AgentEpsilon) [reply browser/components/MarkdownPreview.js]: Yes! Added `printout.destroy()` regardless of success or failure.

**Comment by AgentEpsilon** [comment]
Sorry for the long wait. Images were not working because I was passing the formatted html through a dataurl. I fixed this through a `<base>` tag in the htmlContentFormatter. I also destroy the render window after the export is finished.

**Comment by MihaelBercic** [comment]
Hello, sorry if this question is not appropriate.

How can i add this to my BoostNote?
I only have .md, .txt and .html possible exports.

Thank you very much.

File: browser/components/MarkdownPreview.js [hunk 2]
```diff
@@ -294,58 +295,77 @@ export default class MarkdownPreview extends React.Component {
...
+            </html>`
+  }
+
+  handleSaveAsHtml () {
+    this.exportAsDocument('html', (noteContent, exportTasks, targetDir) => Promise.resolve(this.htmlContentFormatter(noteContent, exportTasks, targetDir)))
+  }
+
+  handleSaveAsPdf () {
+    this.exportAsDocument('pdf', (noteContent, exportTasks, targetDir) => {
+      const printout = new remote.BrowserWindow({show: true, webPreferences: {webSecurity: false}})
```
Comment (ZeroX-DG) [inline browser/components/MarkdownPreview.js]: Just a small request, can you change the browser window `show` property to false? Thanks!

**Comment by ZeroX-DG** [comment]
@MihaelBercic It's still in progress and hasn't been merged into the main project so you have to wait for that. Sorry for the inconvenience.

File: browser/components/MarkdownPreview.js (hunk 2 above)
Comment (AgentEpsilon) [reply browser/components/MarkdownPreview.js]: Done! This was set to `true` for debugging purposes - looks like I just forgot to switch it back.

**Comment by MihaelBercic** [comment]
Thank you.
I’ll wait :)

**Comment by masa4u** [comment]
export note to pdf not working on the windows version.

 * OS: Windows 7

error message:
 Export error: Callback must be a function

**Pull request merged**